README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

# pbp_django_auth_extended

Forked from pbp_django_auth.
A Flutter package for helping students to implement authentication from Django web service in Flutter.
This forked (extended) package adds features that is not present in pbp_django_auth packages.

<!--## Features

List what your package can do. Maybe include images, gifs, or videos.-->

## Getting Started

### Django's Part

To use the package, you need to make asynchronous JavaScript (AJAX) login view in your Django project.

1. Run `python manage.py startapp authentication` to make a new app module for handling the AJAX login.
2. Add `"authentication"` to `INSTALLED_APPS` in `settings.py`.
3. Run `pip install django-cors-headers` to install the required library.
4. Add `"corsheaders"` to `INSTALLED_APPS` in `settings.py`.
5. Add `"corsheaders.middleware.CorsMiddleware"` to `MIDDLEWARE` in `settings.py`.
6. Create a new variable in `settings.py` called `CORS_ALLOW_ALL_ORIGINS` and set the value to `True` (`CORS_ALLOW_ALL_ORIGINS=True`).
7. Create a new variable in `settings.py` called `CORS_ALLOW_CREDENTIALS` and set the value to `True`, (`CORS_ALLOW_CREDENTIALS=True`).
8. Create the following variables in `settings.py`.

    ```python
    CSRF_COOKIE_SECURE = True
    SESSION_COOKIE_SECURE = True
    CSRF_COOKIE_SAMESITE = 'None'
    SESSION_COOKIE_SAMESITE = 'None'
    ```

9. Create a login view method in `authentication/views.py`.
  
    **Example Login View**

    ```python
    from django.shortcuts import render
    from django.contrib.auth import authenticate, login as auth_login
    from django.http import JsonResponse
    from django.views.decorators.csrf import csrf_exempt

    @csrf_exempt
    def login(request):
        username = request.POST['username']
        password = request.POST['password']
        user = authenticate(username=username, password=password)
        if user is not None:
            if user.is_active:
                auth_login(request, user)
                # Redirect to a success page.
                return JsonResponse({
                  "status": True,
                  "message": "Successfully Logged In!"
                  # Insert any extra data if you want to pass data to Flutter
                }, status=200)
            else:
                return JsonResponse({
                  "status": False,
                  "message": "Failed to Login, Account Disabled."
                }, status=401)

        else:
            return JsonResponse({
              "status": False,
              "message": "Failed to Login, check your email/password."
            }, status=401)
    ```

    This view will set cookies to the user and allow authenticated requests with `@login_required` decorator.

10. Create a check_is_anonymous view method in `authentication/views.py`.

    ```python
    def check_is_anonymous(request):
      print(request.user.is_anonymous)
      return JsonResponse({
          "anonymous": request.user.is_anonymous
      }, status=200)
    ```

11. Add url to `authentication/urls.py`

    ```python
    path('is-anonymous/', check_is_anonymous, name='check_is_anonymous'),
    ```

### Flutter's Part

To use the package, modify application root widget to provide the `CookieRequest` library to all child widgets by using `Provider`.

For example, if the previous app initialization was:

```dart
class MyApp extends StatelessWidget {
    const MyApp({Key? key}) : super(key: key);
    
    @override
    Widget build(BuildContext context) {
        return MaterialApp(
            title: 'Flutter App',
            theme: ThemeData(
                primarySwatch: Colors.blue,
            ),
            home: const MyHomePage(title: 'Flutter App'),
            routes: {
                "/login": (BuildContext context) => const LoginPage(),
            },
        );
    }
}
```

Change it to:

```dart
class MyApp extends StatelessWidget {
    const MyApp({Key? key}) : super(key: key);

    @override
    Widget build(BuildContext context) {
        return Provider(
            create: (_) {
                CookieRequest request = CookieRequest(baseUrl: <DJANGO BASE URL>); // DO NOT use trailing '/' here
                return request;
            },
            child: MaterialApp(
                title: 'Flutter App',
                theme: ThemeData(
                    primarySwatch: Colors.blue,
                ),
                home: const MyHomePage(title: 'Flutter App'),
                routes: {
                    "/login": (BuildContext context) => const LoginPage(),
                },
            ),
        );
    }
}
```

This creates a new `Provider` object that will share the `CookieRequest` instance with all components in the application.

## Usage

To use the package in your project, follow these steps below.

1. Import the `Provider` library and this package to the component.

    ```dart
    import 'package:provider/provider.dart';
    import 'package:pbp_django_auth/pbp_django_auth.dart';
    ...
    ```

2. Instantiate the `request` object by calling `context.watch` in the Widget `build(BuildContext context)` function.

    **Example**

    ```dart
    class _LoginPageState extends State<LoginPage> {
      final _loginFormKey = GlobalKey<FormState>();
      bool isPasswordVisible = false;
      void togglePasswordView() {
        setState(() {
          isPasswordVisible = !isPasswordVisible;
        });
      }

      String username = "";
      String password1 = "";
      @override
      Widget build(BuildContext context) {
        final request = context.watch<CookieRequest>();
        // The rest of your widgets are down below
        ...
      }
    }
    ```

3. To log in using the package, use the `request.login(url, data)` method.

    ```dart
      // 'username' and 'password' should be the values of the user login form.
      final response = await request.login("/auth/login", {
        'username': username,
        'password': password1,
      });
      if (request.loggedIn) {
        // Code here will run if the login succeeded.
      } else {
        // Code here will run if the login failed (wrong username/password).
      }
    ```

4. To fetch or insert data using the library, use the `request.get(url)` or `request.post(url, data)` method.

    ```dart
    /* GET request example: */
    final response = await request.get(<PATH TO ACCESS (DO NOT INCLUDE BASE URL)>);
    // The returned response will be a Map object with the keys of the JsonResponse
    
    /* POST request example: */
    final response = await request.post(<PATH TO ACCESS (DO NOT INCLUDE BASE URL)>, {
      "data1": "THIS IS EXAMPLE DATA",
      "data2": "THIS IS EXAMPLE DATA 2",
    });
    // The data argument should be the keys of the Django form.
    // The returned response will be a Map obejct with the keys of JsonResponse.
    ```

    You can also use request.postJson(url, encodedJsonData) with jsonEncode function from 'dart:convert' library to send the submitted data without manually converting the data into JSON format one by one.

## Contributors

- [Adrian Ardizza](https://github.com/Meta502)
- [Muhammad Athallah](https://github.com/determinedguy)
- [Sabyna Maharani](https://github.com/sabynn)
- [Rendy](https://github.com/rorre)
- [Ahmad Fatih Faizi](https://github.com/cnotgate)

<!--
https://blog.logrocket.com/how-to-create-dart-packages-for-flutter/
https://gist.github.com/Meta502/1605fdba3b141fbf67dba689e9e55498
-->


CHANGELOG.md
## 0.5.3

- Update documentation.

## 0.5.2

- QOL update.

## 0.5.1

- Fix bug causing double concatenation on fetch.

## 0.5.0

- Add ability to provide baseUrl on create.
- All requests now will use baseUrl as suffix to request URL.

## 0.4.1

- Supports checking if user is AnonymousUser in init.

## 0.3.3

- Change http version requirement.

## 0.3.2

- Fork from pbp_django_auth.

## 0.3.1

- Migrate repository to PBP Fasilkom UI organisation.

## 0.3.0

- Add ability to skip expired cookies.
- Fix logout procedure to carry out headers used before.

## 0.2.3

- Fix updateCookie method.

## 0.2.2

- Fix data storage variable type.

## 0.2.1

- Fix format.

## 0.2.0

- Fix LateInitializationError issue.
- Removal of BuildContext as a required parameter.

## 0.1.2

- Add jsonData to facilitate extra data given from Django.

## 0.1.1

- Fix documentation.

## 0.1.0

- Initial release.


lib/pbp_django_auth.dart
library pbp_django_auth;

import 'package:http/http.dart' as http;
import 'dart:convert';
import 'package:flutter/foundation.dart' show kIsWeb;
import 'package:shared_preferences/shared_preferences.dart';

class Cookie {
  String name;
  String value;
  int? expireTimestamp;

  Cookie(this.name, this.value, this.expireTimestamp);

  Cookie.fromJson(Map<String, dynamic> json)
      : name = json['name'],
        value = json['value'],
        expireTimestamp = json['expireTimestamp'];

  Map toJson() => {
        "name": name,
        "value": value,
        "expireTimestamp": expireTimestamp,
      };
}

class CookieRequest {
  Map<String, String> headers = {};
  Map<String, Cookie> cookies = {};
  Map<String, dynamic> jsonData = {};
  final http.Client _client = http.Client();

  late SharedPreferences local;

  bool loggedIn = false;
  bool initialized = false;

  Future init() async {
    if (!initialized) {
      local = await SharedPreferences.getInstance();
      cookies = _loadSharedPrefs();
      if (cookies['sessionid'] != null) {
        loggedIn = true;
        headers['cookie'] = _generateCookieHeader();
      }
    }
    initialized = true;
  }

  Map<String, Cookie> _loadSharedPrefs() {
    String? savedCookies = local.getString("cookies");
    if (savedCookies == null) {
      return {};
    }

    Map<String, Cookie> convCookies = {};

    try {
      var localCookies =
          Map<String, Map<String, dynamic>>.from(json.decode(savedCookies));
      for (String keyName in localCookies.keys) {
        convCookies[keyName] = Cookie.fromJson(localCookies[keyName]!);
      }
    } catch (_) {
      // We do not care if the cookie is invalid, just ignore it
    }

    return convCookies;
  }

  Future persist(String cookies) async {
    local.setString("cookies", cookies);
  }

  Future<dynamic> login(String url, dynamic data) async {
    await init();
    if (kIsWeb) {
      dynamic c = _client;
      c.withCredentials = true;
    }

    http.Response response =
        await _client.post(Uri.parse(url), body: data, headers: headers);

    await _updateCookie(response);

    if (response.statusCode == 200) {
      loggedIn = true;
      jsonData = json.decode(response.body);
    } else {
      loggedIn = false;
    }

    return json.decode(response.body);
  }

  Map<String, dynamic> getJsonData() {
    return jsonData;
  }

  Future<dynamic> get(String url) async {
    await init();
    if (kIsWeb) {
      dynamic c = _client;
      c.withCredentials = true;
    }

    http.Response response =
        await _client.get(Uri.parse(url), headers: headers);
    await _updateCookie(response);

    return json.decode(response.body);
  }

  Future<dynamic> post(String url, dynamic data) async {
    await init();
    if (kIsWeb) {
      dynamic c = _client;
      c.withCredentials = true;
    }

    http.Response response =
        await _client.post(Uri.parse(url), body: data, headers: headers);
    await _updateCookie(response);

    return json.decode(response.body);
  }

  Future<dynamic> postJson(String url, dynamic data) async {
    await init();
    if (kIsWeb) {
      dynamic c = _client;
      c.withCredentials = true;
    }

    // Add additional header
    headers['Content-Type'] = 'application/json; charset=UTF-8';
    http.Response response =
        await _client.post(Uri.parse(url), body: data, headers: headers);

    // Remove used additional header
    headers.remove('Content-Type');
    await _updateCookie(response);

    return json.decode(response.body);
  }

  Future _updateCookie(http.Response response) async {
    await init();

    String? allSetCookie = response.headers['set-cookie'];

    if (allSetCookie != null) {
      // Hacky way to simply ignore expires
      allSetCookie = allSetCookie.replaceAll(
        RegExp(r'expires=.+?;', caseSensitive: false),
        "",
      );
      var setCookies = allSetCookie.split(',');

      for (var cookie in setCookies) {
        _setCookie(cookie);
      }

      headers['cookie'] = _generateCookieHeader();
      String cookieObject = (const JsonEncoder()).convert(cookies);
      persist(cookieObject);
    }
  }

  void _setCookie(String rawCookie) {
    if (rawCookie.isEmpty) {
      return;
    }

    var cookieProps = rawCookie.split(";");

    // First part of cookie will always be the key-value pair
    var keyValue = cookieProps[0].split('=');
    if (keyValue.length != 2) {
      return;
    }

    String cookieName = keyValue[0].trim();
    String cookieValue = keyValue[1];

    int? cookieExpire;
    // Iterate through every props and find max-age
    // Expires works but Django always returns max-age, and according to MDN
    // max-age has higher prio

    for (var props in cookieProps.sublist(1)) {
      var keyval = props.split("=");
      if (keyval.length != 2) {
        continue;
      }

      var key = keyval[0].trim().toLowerCase();
      if (key != 'max-age') {
        continue;
      }

      int? deltaTime = int.tryParse(keyval[1]);
      if (deltaTime != null) {
        cookieExpire = DateTime.now().millisecondsSinceEpoch ~/ 1000;
        cookieExpire += deltaTime;
      }
      break;
    }
    cookies[cookieName] = Cookie(cookieValue, cookieValue, cookieExpire);
  }

  String _generateCookieHeader() {
    int currTime = DateTime.now().millisecondsSinceEpoch ~/ 1000;
    String cookie = "";

    for (var key in cookies.keys) {
      if (cookie.isNotEmpty) cookie += ";";
      Cookie? curr = cookies[key];

      if (curr == null) continue;
      if (curr.expireTimestamp != null && currTime >= curr.expireTimestamp!) {
        if (curr.name == "sessionid") {
          // Reset all states if sessionId got changed
          loggedIn = false;
          jsonData = {};
          cookies = {};
        }
        continue;
      }

      String newCookie = curr.value;
      cookie += '$key=$newCookie';
    }

    return cookie;
  }

  Future<dynamic> logout(String url) async {
    await init();
    if (kIsWeb) {
      dynamic c = _client;
      c.withCredentials = true;
    }

    http.Response response =
        await _client.post(Uri.parse(url), headers: headers);

    if (response.statusCode == 200) {
      loggedIn = false;
      jsonData = {};
    } else {
      loggedIn = true;
    }

    cookies = {};

    return json.decode(response.body);
  }
}


lib/pbp_django_auth_extended.dart
library pbp_django_auth_extended;

import 'package:http/http.dart' as http;
import 'dart:convert';
import 'package:flutter/foundation.dart' show kIsWeb;
import 'package:shared_preferences/shared_preferences.dart';

class Cookie {
  String name;
  String value;
  int? expireTimestamp;

  Cookie(this.name, this.value, this.expireTimestamp);

  Cookie.fromJson(Map<String, dynamic> json)
      : name = json['name'],
        value = json['value'],
        expireTimestamp = json['expireTimestamp'];

  Map toJson() => {
        "name": name,
        "value": value,
        "expireTimestamp": expireTimestamp,
      };
}

class CookieRequest {
  CookieRequest({required this.baseUrl});

  Map<String, String> headers = {};
  Map<String, Cookie> cookies = {};
  Map<String, dynamic> jsonData = {};
  final http.Client _client = http.Client();

  final String baseUrl;

  late SharedPreferences local;

  bool loggedIn = false;
  bool initialized = false;

  Future init() async {
    if (!initialized) {
      local = await SharedPreferences.getInstance();
      cookies = _loadSharedPrefs();
      if (cookies['sessionid'] != null) {
        headers['cookie'] = _generateCookieHeader();
      }
      initialized = true;
      var response = await get("/authentication/is-anonymous/");
      if (response['anonymous'] == false) {
        loggedIn = true;
      }
    }
    initialized = true;
  }

  Map<String, Cookie> _loadSharedPrefs() {
    String? savedCookies = local.getString("cookies");
    if (savedCookies == null) {
      return {};
    }

    Map<String, Cookie> convCookies = {};

    try {
      var localCookies =
          Map<String, Map<String, dynamic>>.from(json.decode(savedCookies));
      for (String keyName in localCookies.keys) {
        convCookies[keyName] = Cookie.fromJson(localCookies[keyName]!);
      }
    } catch (_) {
      // We do not care if the cookie is invalid, just ignore it
    }

    return convCookies;
  }

  Future persist(String cookies) async {
    local.setString("cookies", cookies);
  }

  Future<dynamic> login(String path, dynamic data) async {
    await init();
    if (kIsWeb) {
      dynamic c = _client;
      c.withCredentials = true;
    }

    http.Response response = await _client.post(Uri.parse(baseUrl + path),
        body: data, headers: headers);

    await _updateCookie(response);

    if (response.statusCode == 200) {
      loggedIn = true;
      jsonData = json.decode(response.body);
    } else {
      loggedIn = false;
    }

    return json.decode(response.body);
  }

  Map<String, dynamic> getJsonData() {
    return jsonData;
  }

  Future<dynamic> get(String path) async {
    await init();

    if (kIsWeb) {
      dynamic c = _client;
      c.withCredentials = true;
    }

    http.Response response =
        await _client.get(Uri.parse(baseUrl + path), headers: headers);
    await _updateCookie(response);

    return json.decode(response.body);
  }

  Future<dynamic> post(String path, dynamic data) async {
    await init();
    if (kIsWeb) {
      dynamic c = _client;
      c.withCredentials = true;
    }

    http.Response response = await _client.post(Uri.parse(baseUrl + path),
        body: data, headers: headers);
    await _updateCookie(response);

    return json.decode(response.body);
  }

  Future<dynamic> postJson(String path, dynamic data) async {
    await init();
    if (kIsWeb) {
      dynamic c = _client;
      c.withCredentials = true;
    }

    // Add additional header
    headers['Content-Type'] = 'application/json; charset=UTF-8';
    http.Response response = await _client.post(Uri.parse(baseUrl + path),
        body: data, headers: headers);

    // Remove used additional header
    headers.remove('Content-Type');
    await _updateCookie(response);

    return json.decode(response.body);
  }

  Future _updateCookie(http.Response response) async {
    await init();

    String? allSetCookie = response.headers['set-cookie'];

    if (allSetCookie != null) {
      // Hacky way to simply ignore expires
      allSetCookie = allSetCookie.replaceAll(
        RegExp(r'expires=.+?;', caseSensitive: false),
        "",
      );
      var setCookies = allSetCookie.split(',');

      for (var cookie in setCookies) {
        _setCookie(cookie);
      }

      headers['cookie'] = _generateCookieHeader();
      String cookieObject = (const JsonEncoder()).convert(cookies);
      persist(cookieObject);
    }
  }

  void _setCookie(String rawCookie) {
    if (rawCookie.isEmpty) {
      return;
    }

    var cookieProps = rawCookie.split(";");

    // First part of cookie will always be the key-value pair
    var keyValue = cookieProps[0].split('=');
    if (keyValue.length != 2) {
      return;
    }

    String cookieName = keyValue[0].trim();
    String cookieValue = keyValue[1];

    int? cookieExpire;
    // Iterate through every props and find max-age
    // Expires works but Django always returns max-age, and according to MDN
    // max-age has higher prio

    for (var props in cookieProps.sublist(1)) {
      var keyval = props.split("=");
      if (keyval.length != 2) {
        continue;
      }

      var key = keyval[0].trim().toLowerCase();
      if (key != 'max-age') {
        continue;
      }

      int? deltaTime = int.tryParse(keyval[1]);
      if (deltaTime != null) {
        cookieExpire = DateTime.now().millisecondsSinceEpoch ~/ 1000;
        cookieExpire += deltaTime;
      }
      break;
    }
    cookies[cookieName] = Cookie(cookieValue, cookieValue, cookieExpire);
  }

  String _generateCookieHeader() {
    int currTime = DateTime.now().millisecondsSinceEpoch ~/ 1000;
    String cookie = "";

    for (var key in cookies.keys) {
      if (cookie.isNotEmpty) cookie += ";";
      Cookie? curr = cookies[key];

      if (curr == null) continue;
      if (curr.expireTimestamp != null && currTime >= curr.expireTimestamp!) {
        if (curr.name == "sessionid") {
          // Reset all states if sessionId got changed
          loggedIn = false;
          jsonData = {};
          cookies = {};
        }
        continue;
      }

      String newCookie = curr.value;
      cookie += '$key=$newCookie';
    }

    return cookie;
  }

  Future<dynamic> logout(String path) async {
    await init();
    if (kIsWeb) {
      dynamic c = _client;
      c.withCredentials = true;
    }

    http.Response response =
        await _client.post(Uri.parse(baseUrl + path), headers: headers);

    if (response.statusCode == 200) {
      loggedIn = false;
      jsonData = {};
    } else {
      loggedIn = true;
    }

    cookies = {};

    return json.decode(response.body);
  }
}
